### PR TITLE
Neo4j stable update

### PR DESCRIPTION
**Pull request by moxious**
This PR does the following:
- Updates to a more recent version of neo4j
- Adds env vars to accept the enterprise license so that containers will start correctly; without these they won't externally.
- Does a bit of cleanup on NOTES so that users get more informative and specific output.

### CODE REVIEW

**Comment by carrodher** [comment]
LGTM but please increase the chart version in `Chart.yaml`

File: stable/neo4j/templates/core-statefulset.yaml [hunk 1]
```diff
@@ -26,6 +26,8 @@ spec:
         env:
           - name: NEO4J_dbms_mode
             value: CORE
+          - name: NEO4J_ACCEPT_LICENSE_AGREEMENT
+            value: "yes"
```
Comment (timstoop) [inline stable/neo4j/templates/core-statefulset.yaml:30]: I think having this default value in here is problematic. I'm not aware of the actual license involved, but it's generally not a good idea to accept the license for all our end users. It's better if this is moved to a variable that defaults to "no", so an end user has to specifically accept the license himself by overriding that parameter. I think that's a safer path.

That variable should then also be added to the README, of course.

File: stable/neo4j/templates/readreplicas-deployment.yaml [hunk 2]
```diff
@@ -24,6 +24,8 @@ spec:
         env:
           - name: NEO4J_dbms_mode
             value: READ_REPLICA
+          - name: NEO4J_ACCEPT_LICENSE_AGREEMENT
+            value: "yes"
```
Comment (timstoop) [inline stable/neo4j/templates/readreplicas-deployment.yaml:28]: Same as my previous comment, make this a Chart variable that defaults to "no", so an end user has to explicitly accept the license.

**Review by timstoop — CHANGES_REQUESTED** [review]

File: stable/neo4j/templates/core-statefulset.yaml (hunk 1 above)
Comment (moxious) [reply stable/neo4j/templates/core-statefulset.yaml:30]: @timstoop I understand the concern you're raising; on the other side if it defaults to no, a user must clone the repo and modify locally, and the chart isn't directly runnable, which is not desirable.

Would it be acceptable if I added a notice front and center in the README that launching from the repo constitutes acceptance of the license, link this portion of the code for anyone who wishes to change it?

If that isn't acceptable, I understand but I may close the PR, as this change is a key one that makes the containers "just work" out of the box.

Comment (timstoop) [reply stable/neo4j/templates/core-statefulset.yaml:30]: No, a user would not have to clone the repo, the whole idea of variables is that they can be easily overridden by a user :) If you give the variable a clear name like `iAcceptTheLicense` (someone else might have a better suggestion here), they can do something like `helm install stable/neo4j --set iAcceptTheLicence=yes` and the variable would be set correctly. I don't know about you, but we generally keep custom config.yaml for each deployment anyway, so we can version those in Git and easily override the defaults. It's a simple change that does not require the kubernetes project to accept the license on behalf of all users.

**Comment by prydonius** [comment]
Is there any reason this is using the enterprise edition of Neo4j? The community edition does not require acceptance of a EULA. I would make this deploy the community edition by default, with the option of deploying the enterprise edition (and acceptance of the EULA via another value).

Also, in general, if the upstream image requires a user to consciously set an environment variable to accept a EULA or something else that requires attention, then the chart should follow the same UX.

**Comment by prydonius** [comment]
cc @mneedham

File: stable/neo4j/templates/core-statefulset.yaml (hunk 1 above)
Comment (ant31) [reply stable/neo4j/templates/core-statefulset.yaml:30]: I'm highly opposed to default it to yes. 
The agreement has to be explicitly accepted by the users.

**Comment by moxious** [comment]
Alright, comments received.  @timstoop gave me enough detail, I know how to approach this, revision will be be pushed.

File: stable/neo4j/templates/readreplicas-deployment.yaml (hunk 2 above)
Comment (dankohn) [reply stable/neo4j/templates/readreplicas-deployment.yaml:28]: I would strongly recommend changing this to:
```yaml
          - name: NEO4J_ACCEPT_GPLv3_LICENSE_AGREEMENT
            value: "no"
```
Some enterprises refuse to use GPLv3. For example, Google treats GPLv3 in their `restricted` class: https://opensource.google.com/docs/thirdparty/licenses/

Helm Charts do not link with other software (like Helm), but it's still advisable to be as transparent as possible as to what the developer is agreeing to.

File: stable/neo4j/templates/core-statefulset.yaml (hunk 1 above)
Comment (gtirloni) [reply stable/neo4j/templates/core-statefulset.yaml]: The neo4j community edition is based on GPLv3 and I've never seen a user have to explicitly accept that. However, this chart is based on the enterprise edition, which carries some restrictions (see [licensing](https://neo4j.com/licensing/)).

File: stable/neo4j/values.yaml [hunk 3]
```diff
@@ -7,7 +7,7 @@ name: "neo4j"
 
 # Specs for the Neo4j docker image
 image: "neo4j"
-imageTag: "3.2.3-enterprise"
+imageTag: "3.3.4-enterprise"
```
Comment (ant31) [inline stable/neo4j/values.yaml:10]: Could it be the CE by default?

File: stable/neo4j/templates/core-statefulset.yaml (hunk 1 above)
Comment (omkensey) [reply stable/neo4j/templates/core-statefulset.yaml:30]: This is the same issue as the Minecraft chart, I believe, and there the default is "no" and if you don't override the default "false" on the `eula` value the chart does not actually deploy the app and complains at you (via NOTES.txt): https://github.com/kubernetes/charts/blob/master/stable/minecraft/templates/NOTES.txt

File: stable/neo4j/values.yaml (hunk 3 above)
Comment (mneedham) [reply stable/neo4j/values.yaml:10]: @ant31 clustering isn't in the CE version so it only makes sense to use Helm with the EE version.

**Comment by moxious** [comment]
Thanks to all reviewers for posting thoughtful comments.  I think I'm going to close this PR because I need to work more with the folks internally at neo4j to work out a few things.  For example,

- License URL of what you're agreeing to if you set a variable that says you agree
- Whether/how to represent the optionality that license agreement only applies to enterprise, and that it's actually GPLv3 for community
- Whether community or enterprise should be the default, considering community isn't eligible for clustering, and the chart is fundamentally set up for that.

The mechanics of making the chart this or that are straightforward, but when I come back with an improvement, I'll want to make some points about these things overall, improve documentation, and have better reasons why the chart ended up a certain way.

**Comment by prydonius** [comment]
Thanks @moxious for taking this into account, and I look forward to the follow up PR. It might be useful for you to take a look at how the Minecraft chart handles the EULA acceptance for the enterprise scenario. By default, the chart will "succeed" to install, but will not [create any Deployments](https://github.com/kubernetes/charts/blob/master/stable/minecraft/templates/deployment.yaml#L1) and will output the following in the [NOTES.txt](https://github.com/kubernetes/charts/blob/master/stable/minecraft/templates/NOTES.txt#L3).

```
##############################################################################
####   ERROR: You did not agree to the EULA in your 'helm install' call.  ####
##############################################################################

This deployment will be incomplete until you read the Minecraft EULA linked
in the README.md, then:

    helm upgrade {{ .Release.Name }} \
        --set minecraftServer.eula=true stable/minecraft
```

If the user accepts the EULA, then simply running the above output command will enable the workload to be deployed and set the required env var.

**Pull request closed without merging**